File: daily_coding_workout/14-nov-2018/Is_BST.js
// function to be implemented
// insert(data)
// remove(data)

// Helper function
// findMinNode()
// getRootNode()
// inorder(node)
// preorder(node)
// postorder(node)
// search(node, data)

class Node {
  constructor(data, left = null, right = null) {
    this.data = data;
    this.right = null;
    this.left = null;
  }
}

class BST {
  constructor() {
    this.root = null;
  }
  insert(data) {
    let newNode = new Node(data);
    if (root === null) {
      root = newNode;
    } else {
      this.insertInBST(this.root, newNode);
    }
  }
  insertInBST(node, newNode) {
    if (newNode.data < node.data) {
      if (node.left === null) {
        node.left = newNode;
      } else {
        this.insertInBST(node.left, newNode);
      }
    } else {
      if (newNode.data > node.data) {
        if (node.right === null) {
          node.right = newNode;
        } else {
          this.insertInBST(node.left, newNode);
        }
      }
    }
  }
}

File: DailyCodingPuzzles/10-nov-2018/isBst.java

class Node {
    int data;
    Node left;
    Node right;

    public Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;

    }
}

public class isBst {
    Node root;

    public isBst() {
        root = null;
    }

    boolean checkBST(Node root) {
        return checkBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean checkBST(Node root, int left, int right) {
        if (root == null)
            return true;
        if (root.data >= right || left >= root.data)
            return false;
        return checkBST(root.left, left, root.data) && checkBST(root.right, root.data, right);
    }
}
